File: Year 3/Parallel and Distributed Programming/ConvolutionMatrix C++/ConvolutionMatrix C++ Parallel/Source Files/Singleton/IOHandler.cpp
#include "../../Header Files/Singleton/IOHandler.h"
#include <fstream>
#include <iostream>

IOHandler* IOHandler::instance = nullptr;

IOHandler* IOHandler::getInstance() {
	if (instance == nullptr) {
		instance = new IOHandler();
	}
	return instance;
}

vector<vector<int>> IOHandler::readMatrixFromFile(const string& fileName) {
	// read the matrix from the file
	vector<vector<int>> matrix;
	ifstream file(fileName);
	if (file.is_open()) {
		int n, m;
		file >> n >> m;
		matrix.resize(n);
		for (int i = 0; i < n; ++i) {
			matrix[i].resize(m);
			for (int j = 0; j < m; ++j) {
				file >> matrix[i][j];
			}
		}
		file.close();
	}
	else {
		cout << "Unable to open file " << fileName << endl;
	}
	return matrix;
}

void IOHandler::writeMatrixToFile(const vector<vector<int>>& matrix, const string& fileName) {
	// write the matrix to the file
	ofstream file(fileName);
	if (file.is_open()) {
		int n = matrix.size();
		int m = matrix[0].size();
		for (int i = 0; i < n; ++i) {
			for (int j = 0; j < m; ++j) {
				file << matrix[i][j] << " ";
			}
			file << endl;
		}
		file.close();
	}
	else {
		cout << "Unable to open file " << fileName << endl;
	}
}

File: Year 3/Parallel and Distributed Programming/ConvolutionMatrix Java/src/Singleton/IOHandler.java
package Singleton;

import java.io.*;
import java.util.Arrays;

public class IOHandler {
    private static IOHandler instance = null;
    private IOHandler() {
    }
    public static IOHandler getInstance() {
        if (instance == null) {
            instance = new IOHandler();
        }
        return instance;
    }

    private String getAbsolutePath(String fileName) {
        File resourceFile = new File(fileName);
        String path = resourceFile.getAbsolutePath();
        if (path.contains("out\\production\\ConvolutionMatrix Java")) {
            path = path.replace("src\\", "");
        }
        return path;
    }

    public int [][] readMatrixFromFile(String fileName) throws IOException{
        int[][] matrix;
        try (BufferedReader br = new BufferedReader(new FileReader(getAbsolutePath(fileName)))) {
            String line = br.readLine();
            String[] dimensions = line.split(" ");
            int N = Integer.parseInt(dimensions[0]);
            int M = Integer.parseInt(dimensions[1]);

            matrix = new int[N][M];

            int row = 0;
            while ((line = br.readLine()) != null) {
                String[] values = line.split(" ");
                for (int col = 0; col < M; col++) {
                    matrix[row][col] = Integer.parseInt(values[col]);
                }
                row++;
            }
        }
        return matrix;
    }

    public void writeMatrixToFile(int[][] matrix, String fileName) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(getAbsolutePath(fileName)))) {
            for (int[] ints : matrix) {
                writer.write(Arrays.toString(ints).replaceAll("[\\[\\],]", ""));
                writer.newLine();
            }
        }
    }
}
